### ISSUE TITLE
Dynamic Variable Groups?

### ISSUE THREAD

**Issue by jlhjr** [MicrosoftDocs/vsts-docs#3702]
Is there a way to reference a variable group dynamically so that it can be selected when the build is queued?  For instance, in the pipeline YAML:

variables:
  - group: '$(VARIABLE_GROUP_NAME)'

The use case is to create a variable group per environment.  I realize another approach would be to create a release pipeline and scope each variable group to a different stage, but for a variety of reasons I prefer a single YAML file, and this seems to be the way Azure DevOps is moving anyway.

Is it supported yet?  Thanks!

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 4827aa99-ae62-bd63-6f2f-a87a4065ed27
* Version Independent ID: c9e461ee-7f42-3503-9460-18fa3a807bbb
* Content: [Variable groups for Azure Pipelines and TFS - Azure Pipelines](https://docs.microsoft.com/en-us/azure/devops/pipelines/library/variable-groups?view=azure-devops&tabs=yaml)
* Content Source: [docs/pipelines/library/variable-groups.md](https://github.com/MicrosoftDocs/vsts-docs/blob/master/docs/pipelines/library/variable-groups.md)
* Product: **devops**
* GitHub Login: @alexhomer1
* Microsoft Alias: **ahomer**

**Comment by BevanG** [comment]
I would like this too. I have multiple build files that are simply passing variables through to templates.

If possible, and I assume it is not, it would be good to extend this to be able to define variables in the include and exclude rules for build triggers. Would result in multiple builds with Variables defined all pointing to a single YAML file.

**Comment by alexhomer1** [comment]
Thank you for your feedback. I have passed this on to the product team for advice.

**Comment by ashokirla** [comment]
@jlhjr @BevanG : This is not supported today. You need to declare the variable group you want to consume in a YAML pipeline. It is the same case with variables in build trigger definition. 

We will look into this feedback. 
Do you expect to change the variable group for every pipeline run or do you intend to maintain a single YAML file and set the variable group value once per pipeline definition using that YAML file.

**Comment by jlhjr** [comment]
Understood, thanks for considering this.  If I've understood your question correctly, ideally I would like the option to set the variable group for each run.  I would still like to use a single YAML file, but I don't want to hardcode the variable group name in that YAML file.

**Comment by danieleusebio** [comment]
I have the same need. Please deploy the feature. Thank you.

**Comment by alexche8** [comment]
Would be helpful to have such functionality. For example for simple pipelines files - I want to have one pipeline file that can be responsible for builds of two branches. And depends on which branch has triggered the build - the associated group of variables can be used for the steps.

Check please my question on stackoverflow. There are syntax errors, but the idea is clean I think https://stackoverflow.com/questions/56224798/can-group-name-variable-be-dynamic-in-azure-pipelines

**Comment by ejjpi** [comment]
This feature would be extremely useful to avoid code and pipelines duplication. NEED.

**Comment by jschrade** [comment]
This is also a feature my team would use daily.

**Comment by yairhdz** [comment]
Same need for our projects.
Please add this feature. It would be very useful! 
Thanks!

**Comment by deploymentking** [comment]
Just to add some more feedback regarding the usefulness of this. look forward to hearing any updates about this

**Comment by cortex93** [comment]
> a single YAML file, and this seems to be the way Azure DevOps is moving anyway.

Does it means build and release pipelines management are going to be merged ?

**Comment by antonpunith** [comment]
I need this too. It is logical to have a group for each environment and use one when pipeline runs.

**Comment by NirajCricket** [comment]
I was running through the same issue.
need to have sets of build running via different sets of variables.
-my solution
 -- create separate VG (variable group)
--create individual Agent Job for every set of build tasks via YAML file, and assign VG per build sets which will be passed  looking the VG, which you need to link in your pipeline by going to triggers>variables > link variable
sample :
jobs:
- job: Linux1
  pool:
    vmImage: 'ubuntu-16.04'
  variables:
  - group: api_vg-prod
  steps:
  - task: FileTransform@1xx

**Comment by mjalas** [comment]
This would be a very useful feature, since could also enable using the strategy matrix for a job to specify the groups to use for different environments. Looking forward to hopefully seeing this feature in the near future.

**Comment by mafiaazul** [comment]
I would love to have this feature as well...

**Comment by Csmithson-qh** [comment]
Throwing in my vote for this feature as well!

**Comment by thafters** [comment]
My team would definitely use this feature.

**Comment by bourquep** [comment]
Does anybody here have an elegant workaround?

**Comment by mjalas** [comment]
The closest workaround I was able to do was using a combination of variables and job templates. The end result was close enough, at least for now, to what dynamic variable groups with strategy matrix would have provided. However, it is not as explicit and far as I understood it required an extra file for the template.

**Comment by davoodharun** [comment]
+1 on this

**Comment by batorskik** [comment]
+1 It would be extremely useful.

**Comment by davoodharun** [comment]
I actually found somewhat of a work around for this if you use a script to hit the azure devops rest api for variable groups. 

https://docs.microsoft.com/en-us/rest/api/azure/devops/distributedtask/variablegroups/get%20variable%20groups?view=azure-devops-rest-5.1

**Comment by jlhjr** [comment]
For what it's worth, our workaround is to create a new build definition per "environment" with the variable values embedded in that definition.

Not elegant by any means, and the DevOps UI doesn't really enable this (you can't "clone" a build definition, for instance), but you can automate some of it through the [REST API](https://docs.microsoft.com/en-us/rest/api/azure/devops/build/definitions/create?view=azure-devops-rest-5.1), as others have pointed out in their solutions.

**Comment by abrignola** [comment]
+1

**Comment by atbagga** [comment]
For those who are blocked on this and using REST-APIs I wrote a util script using Azure-Devops CLI extension to achieve this. You can take a look at it here - https://github.com/atbagga/PsSamples/tree/master/DevopsVarGroup

This will make it a little easier to consume in your automation.

**Comment by x10an14** [comment]
+1 We would also be in favor of the utmost celerity (hehe) in the addition of this feature! =)

**Comment by UmairSyed** [comment]
+1

**Comment by winstonhenke** [comment]
+1 on this

**Comment by mortenlerudjordet** [comment]
You can also set a shared variable using powershell and the vsTeam module.
Did a small write-up here:
http://blog.lerun.info/2019/09/28/azure-devops-pass-variable-from-build-to-release/

**Comment by x10an14** [comment]
@mortenlerudjordet 
Awesome =) Unfortunately I myself (and maybe others too) find the `azure-pipelines.yaml` pipelines-as-code feature to be more comfortable/fitting with our workflows^^

Additionally, in your write-up, I didn't catch how you would handle identical variable names with differing values depending on environment you deploy to?

**Comment by mortenlerudjordet** [comment]
@x10an14, I too would prefer to be able to both target -group by using a variable name $(VariableGroupName), and be able to update/create shared group variables in AzP YAML. 

I believe identical variable names are not supported inside the same shared variable group, though one could have different variable groups that has the same named variables. The issue here is that the Release pipeline does not support it. Would be nice if the Release pipeline could understand something like: $(VariableGroupName\VariableName) to select from different linked variable groups. 

One workaround to this could be to use the same code in Build for interacting with variable groups, but now read instead. Would jeg need to have a robust way of passing in variable names, and setting local ones that the pipeline can use in other tasks.

**Comment by kerenkeren** [comment]
I would love to have this feature as well...

**Comment by mboutet** [comment]
+1 we also need this.

**Comment by nanderson-gy** [comment]
+1 we need this as well

**Comment by WimDiepgrond** [comment]
This would save a lot of duplication in pipelines +1

**Comment by CousinOfDeath** [comment]
+1 we also need this.

**Comment by JG127** [comment]
Same for me. Because I'd like to have a single azure-pipelines.yml file people can reuse as is.

**Comment by beasamu1** [comment]
+1 on this!

**Comment by Bhanu88** [comment]
+1 for the mentioned feature

**Comment by ssorokoski** [comment]
+1 yes

**Comment by fd-emil** [comment]
+1 please !

**Comment by NateB2** [comment]
We could also really use this feature.

**Comment by ozgunakalin** [comment]
+1 here

**Comment by KKnapik** [comment]
:) 1 up

**Comment by jedthornock** [comment]
+1 here too!

**Comment by jeganbaskar** [comment]
+ 1 its been a long time the request is pending any road map on the request atleast

**Comment by enamba** [comment]
+1 here !! pls

**Comment by michwii** [comment]
Hi guys,

Here is how I have selected dynamically the name of my variable group : 
```
variables:
- ${{ if eq(variables['build.SourceBranchName'], 'prod') }}:
  - group: prod
- ${{ if eq(variables['build.SourceBranchName'], 'staging') }}:
  - group: staging
- ${{ if not(and(eq(variables['build.SourceBranchName'], 'staging'), eq(variables['build.SourceBranchName'], 'prod'))) }}:
  - group: test 
```

Hope it will help you.

Regards

**Comment by lzhwxy** [comment]
michwii, 
    Your syntax doesn't work for me, have you tested it?

Thanks

**Comment by lzhwxy** [comment]
+1
have been looking for solution for the whole day... we do need this feature if all possible.

**Comment by svaenn** [comment]
+1!

**Comment by ppawiggers** [comment]
+1

**Comment by ffalorjr** [comment]
+1 I have a highly reusable pipeline template that relies on variable groups and pipeline variables to make maintaining easier. Gonna try @michwii work around and base my IF statements on the pipeline name. This should give me what I want in the meantime.

**Comment by MikaMaroqui** [comment]
+1

**Comment by JesseStolwijk** [comment]
+1

**Comment by Marco10101** [comment]
+1

**Comment by MikeSzulik** [comment]
+1

**Comment by dushyantovyas** [comment]
+1

**Comment by wolszakp** [comment]
Any news in that topic?
It would be also useful in my setup. 
I really like feature to have most variables in repository. 
And only few of them assigned dynamically from DevOps that will control whole solution.

**Comment by svaenn** [comment]
@jlhjr I posted your question as a suggestion [here](https://developercommunity.visualstudio.com/idea/905525/reference-a-variable-group-dynamically-by-queue-ti.html)
@all: Maybe this will get some more attention if we vote it up there.

**Comment by mszadziul87** [comment]
+1 on this!

**Comment by pmbrull** [comment]
+1

**Comment by motowilliams** [comment]
This is a fairly basic thing that, in this case, sets Octopus Deploy apart from this. 

This needs to be implemented.

**Comment by worldspawn** [comment]
Hi folks, just wanted to share my revolting solution for this.

Define username and password variables for your pipeline. Password should be a devops token with read access to variables.

``` yaml
- job: DefineVariables
      steps:
        - checkout: none
        - pwsh: |
            $region = "";
            $env = ""
            $workspace = ""
            
            switch ("$(envtarget)") {
              "test" {
                $region = "au"
                $env = "dev"
                $workspace = "test"
                break;
              }
              "dev" {
                $region = "us"
                $env = "dev"
                $workspace = "dev"
                break;
              }
            }

            $username = $env:USERNAME
            $password = "$(password)"

            $groupName = "processing_$($region)_$($env)_$($workspace)"

            $auth = [System.Convert]::ToBase64String([System.Text.Encoding]::UTF8.GetBytes("$($userName):$password"))
            $variableGroups = Invoke-WebRequest -Uri "https://dev.azure.com/yourtenant/yourproject/_apis/distributedtask/variablegroups?api-version=5.1-preview.1" -Headers @{"Authorization"="Basic $auth"} -UseBasicParsing | ConvertFrom-Json
            $input = ($variableGroups.value | ? { $_.name -eq $groupName } | Select-Object -first 1).id
            if ($input -eq $null) { exit 1 }
            $group = Invoke-WebRequest -Uri "https://dev.azure.com/yourtenant/yourproject/_apis/distributedtask/variablegroups/$($input)?api-version=5.1-preview.1" -Headers @{"Authorization"="Basic $auth"} -UseBasicParsing | ConvertFrom-Json
            $variables = $group.variables

            $variables.psobject.properties | %{ 
              Write-Host "##vso[task.setvariable variable=$($_.name);isOutput=true]$($_.value.value)"
            }

            Write-Host "##vso[task.setvariable variable=region;isOutput=true]$region"
            Write-Host "##vso[task.setvariable variable=env;isOutput=true]$env"
            Write-Host "##vso[task.setvariable variable=workspace;isOutput=true]$workspace"
```

Sub out `yourtenant/yourproject` as appropriate.

**Comment by worldspawn** [comment]
Of course when you try to use it  it doesn't work

```
variables:
        registry: $[ dependencies.DefineVariables.outputs['setvars.registry']]
```
```
- task: Docker@2
        displayName: Login to ACR
        inputs:
          command: login
          #containerRegistry: "drawboard$[variables.region]-registry"
          #containerRegistry: $(registry)
          containerRegistry: $[variables.registry]
```
>Step Docker1 input containerRegistry references service connection $[variables.registry] which could not be found.

Truly awful product.

**Comment by vtbassmatt** [comment]
We're introducing [runtime parameters](https://github.com/microsoft/azure-pipelines-yaml/blob/master/design/runtime-parameters.md) next sprint (ETA is early March) which will solve most of these use cases.

Also, for future reference, product feedback should be registered with [Developer Community](https://developercommunity.visualstudio.com/spaces/21/index.html). That's the area we check for requested new features, not the docs repo.

**Comment by NickSpag** [comment]
If you're just looking to change based on environments, and your environment can be resolved via the built in conditionals, this is valid yaml:

```
variables:
- ${{ if eq(variables['Build.SourceBranch'], 'refs/heads/master') }}:
  - group: 'AzureDevopsLibraryName-PROD'
```

There's probably a lot more you can do with this while we await the runtime parameters that are coming, but as an example to resolve different environments for a git flow style repo:
```
variables:
  - name: 'AppSettingsPath'
    value: '$(System.DefaultWorkingDirectory)/functions/MyFunction/' 
- ${{ if or(eq(variables['Build.SourceBranch'], 'refs/heads/develop'), startswith(variables['Build.SourceBranch'], 'refs/heads/feature'), startswith(variables['Build.SourceBranch'], 'refs/heads/chore'), startswith(variables['Build.SourceBranch'], 'refs/heads/hotfix')) }}:
  - group: 'AzureDevopsLibraryName-DEV'
- ${{ if startswith(variables['Build.SourceBranch'], 'refs/heads/release') }}:
  - group: 'AzureDevopsLibraryName-QA'
- ${{ if eq(variables['Build.SourceBranch'], 'refs/heads/master') }}:
  - group: 'AzureDevopsLibraryName-PROD'
```

**Comment by Planche95** [comment]
Anybody knows how can I dynamically assign variable groups for release pipelines? not YAML build pipelines

**Comment by vtbassmatt** [comment]
@Planche95 I don't believe that is supported in classic RM.

**Comment by davejhahn** [comment]
Not what is being asked, definitely a workaround, but you could store variables in an environment specific keyvault, and based on something (e.g. a passed in variable or looking at a system variable), within a PowerShell task, set a KeyVault variable, then have a task after that to pull from the Key Vault.  Provided the variable names are the same for each environment.

e.g.
1. PowerShell: if "development" (however you are determining that), set KeyVault variable name to use to Dev keyvault
2. Azure KeyVault Task, loading from above keyvault varaible.  Variable names are the keyvault keys.
3. Use keys.

Obviously that means storing the values in keyvault.  For my use case, it's just a small subset from the full environment variable group so it's a reasonable workaround.

**Comment by DotNetWebAppDeveloper** [comment]
We are trying to reference a variable group dynamically with if condition. We get a error message 
![image](https://user-images.githubusercontent.com/19928575/80582521-277f0600-8a0f-11ea-8641-c120f9ac25f9.png)
**Error message:**
![image](https://user-images.githubusercontent.com/19928575/80582533-2e0d7d80-8a0f-11ea-96af-cfb2bf4a0194.png)
There are other new syntax on Azure-DevOps?

**Comment by michwii** [comment]
> We are trying to reference a variable group dynamically with if condition. We get a error message
> ![image](https://user-images.githubusercontent.com/19928575/80582521-277f0600-8a0f-11ea-8641-c120f9ac25f9.png)
> **Error message:**
> ![image](https://user-images.githubusercontent.com/19928575/80582533-2e0d7d80-8a0f-11ea-96af-cfb2bf4a0194.png)
> Is there a different or new syntax to write in Azure DevOps?

Try the post I have written below :

> Hi guys,
> 
> Here is how I have selected dynamically the name of my variable group :
> 
> ```
> variables:
> - ${{ if eq(variables['build.SourceBranchName'], 'prod') }}:
>   - group: prod
> - ${{ if eq(variables['build.SourceBranchName'], 'staging') }}:
>   - group: staging
> - ${{ if not(and(eq(variables['build.SourceBranchName'], 'staging'), eq(variables['build.SourceBranchName'], 'prod'))) }}:
>   - group: test 
> ```
> 
> Hope it will help you.
> 
> Regards

Don't worry about the syntax error highlighting, I also have it in my editor.

Wish it will help you :)

**Comment by DotNetWebAppDeveloper** [comment]
> > We are trying to reference a variable group dynamically with if condition. We get a error message

> > Is there a different or new syntax to write in Azure DevOps?
> 
> Try the post I have written below :
> 
> > Hi guys,
> > Here is how I have selected dynamically the name of my variable group :
> > ```
> > variables:
> > - ${{ if eq(variables['build.SourceBranchName'], 'prod') }}:
> >   - group: prod
> > - ${{ if eq(variables['build.SourceBranchName'], 'staging') }}:
> >   - group: staging
> > - ${{ if not(and(eq(variables['build.SourceBranchName'], 'staging'), eq(variables['build.SourceBranchName'], 'prod'))) }}:
> >   - group: test 
> > ```
> > 
> > 
> > Hope it will help you.
> > Regards
> 
> Don't worry about the syntax error highlighting, I also have it in my editor.
> 
> Wish it will help you :)

**It works**
Thanks

**Comment by zhangpengchen** [comment]
> > > We are trying to reference a variable group dynamically with if condition. We get a error message
> 
> > > Is there a different or new syntax to write in Azure DevOps?
> > 
> > 
> > Try the post I have written below :
> > > Hi guys,
> > > Here is how I have selected dynamically the name of my variable group :
> > > ```
> > > variables:
> > > - ${{ if eq(variables['build.SourceBranchName'], 'prod') }}:
> > >   - group: prod
> > > - ${{ if eq(variables['build.SourceBranchName'], 'staging') }}:
> > >   - group: staging
> > > - ${{ if not(and(eq(variables['build.SourceBranchName'], 'staging'), eq(variables['build.SourceBranchName'], 'prod'))) }}:
> > >   - group: test 
> > > ```
> > > 
> > > 
> > > Hope it will help you.
> > > Regards
> > 
> > 
> > Don't worry about the syntax error highlighting, I also have it in my editor.
> > Wish it will help you :)
> 
> **It works**
> Thanks

When you have many variables, by using this setup you would like to have sub groups in place which is not possible at this moment.

**Comment by cayodiebe** [comment]
+1!

**Comment by seki-takuya-fixer** [comment]
@zhangpengchen 

```
variables:
- group: ${{ variables['build.SourceBranchName'] }}
```

**Comment by Win32Sector** [comment]
It's really strange. I try this syntax:

variables:
- ${{ if eq(variables['EnvName'], 'prod') }}:
  - group: prod
- ${{ if eq(variables['EnvName'], 'staging') }}:
  - group: staging

EnvName - it's pipeline variable, that define's when I run pipeline, but variable group doesn't take. 
Or I can use predefined variables only if I want to take variable groups dynamically?

**Comment by ALIENQuake** [comment]
@michwii Is there any way to use `Build.Repository.Name` variable for your solution?

**Comment by michwii** [comment]
Hi @ALIENQuake, I never tried but in theory you can.

**Comment by gparissis** [comment]
> It's really strange. I try this syntax:
> 
> ```
> variables:
> - ${{ if eq(variables['EnvName'], 'prod') }}:
>   - group: prod
> - ${{ if eq(variables['EnvName'], 'staging') }}:
>   - group: staging
> ```
> 
> EnvName - it's pipeline variable, that define's when I run pipeline, but variable group doesn't take. Or I can use predefined variables only if I want to take variable groups dynamically?
> 
> But I have lod version of Azure DevOps Server Version Dev17.M153.5 May be it doesn't work on it.

I wanted to do the same but it is not working. Based on my understanding the expression used is a compile time expression, meaning that it is evaluated before the variable is set. Unfortunately runtime expressions are enabled only for variable assignments and stage, job, task condition checks.

**Comment by DidierHoarau** [comment]
> We're introducing [runtime parameters](https://github.com/microsoft/azure-pipelines-yaml/blob/master/design/runtime-parameters.md) next sprint (ETA is early March) which will solve most of these use cases.
> 
> Also, for future reference, product feedback should be registered with [Developer Community](https://developercommunity.visualstudio.com/spaces/21/index.html). That's the area we check for requested new features, not the docs repo.

Thank you for the information. However, I've just check the documentation about Runtime Parameters and I don't really see how it solves the problem that is discussed in this Thread. Am I missing something here?

Thanks.

**Comment by WilliamAntonRohm** [comment]
@All -- thank you for your follow-up.  Please consider these resources:
- [Azure DevOps Services](https://azure.microsoft.com/support/devops/) 
- [Azure DevOps on Stack Overflow](https://stackoverflow.com/questions/tagged/azure-devops)

To request a new feature, [please add it here](https://developercommunity.visualstudio.com/search?space=21).

**Comment by BevanG** [comment]
> It's really strange. I try this syntax:
> 
> ```
> variables:
> - ${{ if eq(variables['EnvName'], 'prod') }}:
>   - group: prod
> - ${{ if eq(variables['EnvName'], 'staging') }}:
>   - group: staging
> ```
> 
> EnvName - it's pipeline variable, that define's when I run pipeline, but variable group doesn't take. Or I can use predefined variables only if I want to take variable groups dynamically?
> 
> But I have lod version of Azure DevOps Server Version Dev17.M153.5 May be it doesn't work on it.

I know this is old know and my first suggestion is already covered, but I'll keep it here for completeness.

@Win32Sector , predefined variables only unfortunately. In your scenario it looks to makes sense to deploy to different environments with different branches anyway. We do this by : 

```
variables:
- ${{ if eq(variables['build.SourceBranchName'], 'develop') }}:
  - group: App.DEV
- ${{ if startsWith(variables['build.SourceBranch'], 'refs/heads/feature/') }}:
  - group: App.Features
- ${{ if startsWith(variables['build.SourceBranch'], 'refs/heads/release/') }}:
  - group: App.Prod
```

If that is not suitable and you really need to use pipeline variables, we get around this with a powershell task to publish them instead. Here is the full code to make it work. @Planche95 this also works for Release pipelines, just need to edit from YAML: 
```
- task: PowerShell@2
  displayName: 'Set Variables for Pipline'
  env:
    AZURE_DEVOPS_EXT_PAT: $(System.AccessToken)
  inputs:
    targetType: 'inline'
    script: |
      # Set default Azure DevOps organization and project
      az devops configure -d organization=https://dev.azure.com/Org-Name project=Application

      # Set Variable group for each environment based on Pull Request target branch
      switch ("$(EnvName)") {
      	# Develop
      	"prod" {
          Write-Host "Publishing Variables from App.Prod"
          Write-Host "-------------------------------------"
      		$Variables = az pipelines variable-group list --group-name App.Prod --project Application | ConvertFrom-Json
      	}
      	# Production
      	"staging" {
          Write-Host "Publishing Variables from App.Staging"
          Write-Host "-------------------------------------"
      		$Variables = az pipelines variable-group list --group-name App.Staging --project Application | ConvertFrom-Json
      	}
      }

      # Pull variables object from Azur DevOps Library into variable
      $Variables = $Variables.Variables

      # Get Variable names from Object
      $Variable_Names = ([regex]::Matches("$Variables",'(([a-z]|[A-Z]).*?)(?==)')).Value

      Foreach ($Variable_Name in $Variable_Names) {
        # Loop through each variable and publish to Pipeline
        Write-Host "$Variable_Name : $($Variables.$($Variable_Name).value)"
      	echo "##vso[task.setvariable variable=$($Variable_Name)]$($Variables.$($Variable_Name).value)"
      }
```

**Comment by chris03** [comment]
> > We're introducing [runtime parameters](https://github.com/microsoft/azure-pipelines-yaml/blob/master/design/runtime-parameters.md) next sprint (ETA is early March) which will solve most of these use cases.
> > Also, for future reference, product feedback should be registered with [Developer Community](https://developercommunity.visualstudio.com/spaces/21/index.html). That's the area we check for requested new features, not the docs repo.
> 
> Thank you for the information. However, I've just check the documentation about Runtime Parameters and I don't really see how it solves the problem that is discussed in this Thread. Am I missing something here?
> 
> Thanks.

You can use a parameter's value as the group name because it get evaluated before the actual pipeline runs.
```
parameters:
  - name: env
    type: string
    values:
      - dev
      - qa
      - prod

variables:
  - group: ${{ parameters.env }}
 ```

**Comment by ashleyghooper** [comment]
> You can use a parameter's value as the group name because it get evaluated before the actual pipeline runs.
> 
> ```
> parameters:
>   - name: env
>     type: string
>     values:
>       - dev
>       - qa
>       - prod
> 
> variables:
>   - group: ${{ parameters.env }}
> ```

I was hoping to be able to pass pre-defined variables like $( Build.Repository.Name) from a 'stub' pipeline YAML that calls a template in another project, but it doesn't work. Is this because the 'pre-defined' variables can not be evaluated until the pipeline is running?

Is there any workaround? I would like to be able to define several variable groups and environments in the same project and have the pipeline dynamically select the correct one to use based on the repo name and branch.

**Comment by del77** [comment]
> > It's really strange. I try this syntax:
> > ```
> > variables:
> > - ${{ if eq(variables['EnvName'], 'prod') }}:
> >   - group: prod
> > - ${{ if eq(variables['EnvName'], 'staging') }}:
> >   - group: staging
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > EnvName - it's pipeline variable, that define's when I run pipeline, but variable group doesn't take. Or I can use predefined variables only if I want to take variable groups dynamically?
> > But I have lod version of Azure DevOps Server Version Dev17.M153.5 May be it doesn't work on it.
> 
> I know this is old know and my first suggestion is already covered, but I'll keep it here for completeness.
> 
> @Win32Sector , predefined variables only unfortunately. In your scenario it looks to makes sense to deploy to different environments with different branches anyway. We do this by :
> 
> ```
> variables:
> - ${{ if eq(variables['build.SourceBranchName'], 'develop') }}:
>   - group: App.DEV
> - ${{ if startsWith(variables['build.SourceBranch'], 'refs/heads/feature/') }}:
>   - group: App.Features
> - ${{ if startsWith(variables['build.SourceBranch'], 'refs/heads/release/') }}:
>   - group: App.Prod
> ```
> 
Also late to the party but I found a solution that was good enough in my case. Instead of using "Source branch name" I used Build.DefinitionName variable which returns the name of the build pipeline. Each of my pipelines has the environment specified in its name so it was enough to pick proper group in the pipeline: 

```
variables:
- ${{ if contains(variables['Build.DefinitionName'], 'testing') }}:
  - group: testing-env-group-name
- ${{ elseif contains(variables['Build.DefinitionName'], 'staging') }}:
  - group: prod-env-group-name
- ${{ elseif contains(variables['Build.DefinitionName'], 'production') }}:
  - group: prod-env-group-name
```

**Comment by htwashere** [comment]
> The closest workaround I was able to do was using a combination of variables and job templates. The end result was close enough, at least for now, to what dynamic variable groups with strategy matrix would have provided. However, it is not as explicit and far as I understood it required an extra file for the template.

I agree, this seems to be the only work around that works end-to-end. The reference https://stackoverflow.com/questions/56224798/can-group-name-variable-be-dynamic-in-azure-pipelines mentioned provided various solutions but it seems in each case, it is only feasible with the value substitution and, by the time you reference the group variable, the pipeline will say that the variable is not found
